src/mds/MDSRankQuiesce.cc

```diff
@@ -0,0 +1,592 @@
+
+  ceph_assert(quiesce_db_manager);
+
+  using RequestHandle = QuiesceInterface::RequestHandle;
+  using QuiescingRoot = std::pair<RequestHandle, Context*>;
+  auto quiesce_requests = std::make_shared<std::unordered_map<QuiesceRoot, QuiescingRoot>>();
+
+  QuiesceAgent::ControlInterface ci;
+
+  ci.submit_request = [this, quiesce_requests](QuiesceRoot root, Context* c)
+      -> std::optional<RequestHandle> {
+    auto uri = boost::urls::parse_uri_reference(root);
+    if (!uri) {
+      dout(5) << "error parsing the quiesce root as an URI: " << uri.error() << dendl;
+      c->complete(uri.error());
+      return std::nullopt;
+    } else {
+      dout(20) << "parsed root '" << root <<"' as : " << uri->path() << " " << uri->query() << dendl;
+    }
+
+    std::chrono::milliseconds quiesce_delay_ms = 0ms;
+    if (auto pit = uri->params().find("delayms"); pit != uri->params().end()) {
+      try {
+        quiesce_delay_ms = std::chrono::milliseconds((*pit).has_value ? std::stoul((*pit).value) : 1000);
+      } catch (...) {
+        dout(5) << "error parsing the time to quiesce for query: " << uri->query() << dendl;
+        c->complete(-EINVAL);
+        return std::nullopt;
+      }
+    }
+    std::optional<double> debug_quiesce_after;
+    if (auto pit = uri->params().find("q"); pit != uri->params().end()) {
+      try {
+        debug_quiesce_after = (*pit).has_value ? std::stod((*pit).value) : 1 /*second*/;
+      } catch (...) {
+        dout(5) << "error parsing the time for debug quiesce for query: " << uri->query() << dendl;
+        c->complete(-EINVAL);
+        return std::nullopt;
+      }
+    }
+    std::optional<double> debug_fail_after;
+    if (auto pit = uri->params().find("f"); pit != uri->params().end()) {
+      try {
+        debug_fail_after = (*pit).has_value ? std::stod((*pit).value) : 1 /*second*/;
+      } catch (...) {
+        dout(5) << "error parsing the time for debug fail for query: " << uri->query() << dendl;
+        c->complete(-EINVAL);
+        return std::nullopt;
+      }
+    }
+    std::optional<mds_rank_t> debug_rank;
+    if (auto pit = uri->params().find("r"); pit != uri->params().end()) {
+      try {
+        if ((*pit).has_value) {
+          debug_rank = (mds_rank_t)std::stoul((*pit).value);
+        }
+      } catch (...) {
+        dout(5) << "error parsing the rank for debug pin for query: " << uri->query() << dendl;
+        c->complete(-EINVAL);
+        return std::nullopt;
+      }
+    }
+
+    if (debug_rank && (debug_rank >= mdsmap->get_max_mds())) {
+        dout(5) << "invalid rank: " << uri->query() << dendl;
+        c->complete(-EINVAL);
+        return std::nullopt;
+    }
+
+    auto path = uri->path();
+    dout(20) << "got request to quiesce '" << path << "'" << dendl;
+
+    std::lock_guard l(mds_lock);
+
+    if (!debug_quiesce_after && !debug_fail_after && !debug_rank) {
+      // the real deal!
+      if (mdsmap->is_degraded()) {
```

Comment:
Why are we doing this check here and not in the DB?

Reply:
Because it's not the DB's job to make conclusions about the MDS state. The DB should be functional at all times, and the decision to fail the quiesce is on the agent's side. It's subject to be revised later whether a degraded cluster should refuse to quiesce. I'll also post a general response to clarify confusion regarding the separation of concerns between the quiesce db and the quiesce agent + protocol.

Reply:
> Because it's not the DB's job to make conclusions about the MDS state. The DB should be functional at all times, and the decision to fail the quiesce is on the agent's side.

It seems like an arbitrary line to draw. Since this is basically an adapter for the QuiesceDB to the quiesce mechanism, it seems odd for it to be making decisions about whether to issue the quiesce request.

> It's subject to be revised later whether a degraded cluster should refuse to quiesce. I'll also post a general response to clarify confusion regarding the separation of concerns between the quiesce db and the quiesce agent + protocol.

I don't see this ever being addressed without wide architectural changes to CephFS. If a rank is down, the file system is basically unavailable. Full stop. It's unproductive to address corner cases.

Reply:
Whatever the limitations are, I would like them to be fully addressed by the agent side. The Quiesce DB is a data structure that should not make assumptions about whether particular roots are quiesceable. The failures can be reported via a standard agent interface, and that's how it's implemented currently.

Reply:
I suppose the reason this is not in `QuiesceAgent.cc` is because that module operates outside of the `mds_lock`?

Reply:
@batrick please resolve this thread if you're satisfied with the explanation